dev/cuda/softmax_forward.cu

```diff
@@ -289,6 +318,30 @@ __global__ void softmax_forward_kernel4(float* out, float* inp, int N, int C) {
     }
 }
 
+__global__ void softmax_forward_online_kernel1(float* out, float* inp, int N, int C) {
+    // inp is (N, C)
+    // out is (N, C), each row of inp will get softmaxed
+    int i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i < N) {
+        float* inp_row = inp + i * C;
+        float* out_row = out + i * C;
+
+        float maxval = -INFINITY;
+        float sum = 0.0f;
+        for (int j = 0; j < C; j++) {
+            float maxval_prev = maxval;
+            if (inp_row[j] > maxval) {
+                maxval = inp_row[j];
+            }
+            sum = sum * expf(maxval_prev - maxval) + expf(inp_row[j] - maxval);
```

Comment:
this value will only be different from 1 if the previous if branch is actually taken, right? So we probably can skip the evaluation of exp most of the time

Reply:
Thanks for the comment, updated the code.